R/runSoupX.R

```diff
@@ -0,0 +1,279 @@
+            newColData$soupX_clusters[sampleIdx[[s]]] <- results[[s]]$sc$metaData$clusters
+            newColData$soupX_contamination[sampleIdx[[s]]] <- results[[s]]$sc$metaData$rho
+            newRowData[[paste0("soupX_",s,"_est")]] <- results[[s]]$sc$soupProfile$est
+            newRowData[[paste0("soupX_",s,"_counts")]] <- results[[s]]$sc$soupProfile$counts
+            S4Vectors::metadata(inSCE)$sctk$SoupX[[s]] <- results[[s]]$sc$fit
+            S4Vectors::metadata(inSCE)$sctk$SoupX[[s]]$nDropUMIs <- results[[s]]$sc$nDropUMIs
+        }
+    }
+    expData(inSCE, assayName, tag = "raw") <- corrAssay
+    SummarizedExperiment::colData(inSCE) <- newColData
+    SummarizedExperiment::rowData(inSCE) <- newRowData
+    
+    return(inSCE)
+}
+
+.SoupXOneBatch <- function(inSCE, 
+                           useAssay, 
+                           background, 
+                           bgAssayName,
+                           cluster, 
+                           tfidfMin, 
+                           soupQuantile,
+                           adjustMethod,
+                           roundToInt,
+                           pCut) {
+    # Building SoupX's SoupChannel object
+    toc <- expData(inSCE, useAssay)
+    if (is.null(background)) {
+        scNoDrops <- SoupX::SoupChannel(toc, toc, calcSoupProfile = FALSE)
+        soupProf <- data.frame(row.names = rownames(toc), 
+                               est = rowSums(toc)/sum(toc), 
+                               counts = rowSums(toc))
+        sc <- SoupX::setSoupProfile(scNoDrops, soupProf)
+    } else {
+        if (inherits(background, "SingleCellExperiment")) {
+            if (is.null(bgAssayName)) {
+                bgAssayName <- useAssay
+            }
+            tod <- SummarizedExperiment::assay(background, bgAssayName)
+        } else {
+            tod <- background
+        }
+        sc <- SoupX::SoupChannel(tod, toc)
+    }
+    
+    # Adding cluster info
+    if (!is.null(cluster)) {
+        if (is.character(cluster) & length(cluster) == 1) {
+            cluster <- SummarizedExperiment::colData(inSCE)[[cluster]]
+            names(cluster) <- colnames(inSCE)
+            sc = SoupX::setClusters(sc, cluster)
+        } else if (length(cluster) == ncol(inSCE)) {
+            names(cluster) <- colnames(inSCE)
+            sc = SoupX::setClusters(sc, cluster)
+        } else {
+            stop("Invalid cluster specification")
+        }
+    } else {
+        message(paste0(date(), " ... Cluster info not given, "))
+        message(paste0(date(), " ...   generating with Scran SNN method."))
+        suppressMessages({
+            inSCE <- scaterlogNormCounts(inSCE, "logcounts", useAssay)
+            inSCE <- runNormalization(inSCE, "logcounts", "scaled", scale = TRUE)
+            inSCE <- seuratFindHVG(inSCE, verbose = FALSE)
+            inSCE <- getTopHVG(inSCE, "vst", altExp = "hvg")
+            inSCE <- scaterPCA(inSCE, "scaled", "hvg")
```

Comment:
what about using the "quickCluster" from scater? Not sure what works better. Does soupx have a recommendation? @mvfki

Reply:
I'll definitely try this out. 
They haven't recommended any method but just saying "The results are not strongly sensitive to the clustering used". But in my own test, clustering in the way I implemented works better than running `runScranSNN(..., useAssay = "counts")`

Reply:
> what about using the "quickCluster" from scater? Not sure what works better. Does soupx have a recommendation? @mvfki

By the way, I just googled it, did you mean "quickCluster" from Scran?

Reply:
yea!